src/inference/ess.jl

```diff
@@ -0,0 +1,132 @@
+"""
+    ESS
+
+Elliptical slice sampling algorithm.
+
+# Examples
+```jldoctest; setup = :(Random.seed!(1))
+julia> @model gdemo(x) = begin
+           m ~ Normal()
+           x ~ Normal(m, 0.5)
+       end
+gdemo (generic function with 2 methods)
+
+julia> sample(gdemo(1.0), ESS(), 1_000) |> mean
+Mean
+
+│ Row │ parameters │ mean     │
+│     │ Symbol     │ Float64  │
+├─────┼────────────┼──────────┤
+│ 1   │ m          │ 0.811555 │
+```
+"""
+struct ESS{space} <: InferenceAlgorithm end
+
+ESS() = ESS{()}()
+ESS(space::Symbol) = ESS{(space,)}()
+
+mutable struct ESSState{V<:VarInfo} <: AbstractSamplerState
+    vi::V
+end
+
+ESSState(model::Model) = ESSState(VarInfo(model))
+
+function Sampler(alg::ESS, model::Model, s::Selector)
+    # sanity check
+    space = getspace(alg)
+    if isempty(space)
+        pvars = get_pvars(model)
+        length(pvars) == 1 ||
+            error("[ESS] no symbol specified to sampler although there is not exactly one model parameter ($pvars)")
+    end
+
+    state = ESSState(model)
+    info = Dict{Symbol, Any}()
+
+    return Sampler(alg, info, s, state)
+end
+
+# always accept in the first step
+function step!(::AbstractRNG, ::Model, spl::Sampler{<:ESS}, ::Integer; kwargs...)
+    return Transition(spl)
+end
+
+function step!(
+    rng::AbstractRNG,
+    model::Model,
+    spl::Sampler{<:ESS},
+    ::Integer,
+    ::Transition;
+    kwargs...
+)
+    # recompute log-likelihood in logp
+    vi = spl.state.vi
+    if spl.selector.tag !== :default
+        runmodel!(model, vi, spl)
```

Comment:
I'm not sure - when should I use `runmodel!(...)` and when just `model(...)`? And do I have to pass `spl` to `runmodel!` here? It's not passed in the MH implementation.

Reply:
I would default to using `runmodel!`, since it resets the `logp` and then runs the model. If your sampler also tracks the number of times the model has been evaluated, `runmodel!` will also update it.

Reply:
Also if you only want to sample from the prior you don't have to pass `spl` to the `runmodel!` call. If you have custom `observe`/`assume` behavior then you should pass the sampler in.

Reply:
OK, thanks! So since I want to get the log-likelihood (and hence have to use my custom assume function, haven't I?) I have to specify `spl`. If I would be interested only in the log joint probability (as, e.g., in the MH algorithm) that would not be necessary, it seems?

Reply:
I believe that's correct, yes.

Reply:
Good :+1: And you don't know of any more idiomatic way to compute the log-likelihood than what I've done in the `assume` function?

Reply:
I think this is as close to idiomatic as it gets in Turing, so you're good.

Reply:
The `LikelihoodContext` is now the idiomatic way in Turing to compute the log likelihood.